File: java/ql/test/utils/modelgenerator/dataflow/p/Sinks.java
package p;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.nio.charset.Charset;
import java.nio.file.CopyOption;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.logging.Logger;

public class Sinks {

  public Object tainted;

  // Defined as a sink in the model file next to the test.
  // neutral=p;Sinks;sink;(Object);summary;df-generated
  public void sink(Object o) {}

  // Defined as a sink in the model file next to the test.
  // neutral=p;Sinks;sink2;(Object);summary;df-generated
  public void sink2(Object o) {}

  // Defined as sink neutral in the file next to the neutral summary test.
  // neutral=p;Sinks;nosink;(Object);summary;df-generated
  public void nosink(Object o) {}

  // Sink and Source defined in the extensible file next to the sink test.
  // sink=p;Sinks;true;saveAndGet;(Object);;Argument[0];test-sink;df-generated
  // neutral=p;Sinks;saveAndGet;(Object);summary;df-generated
  public Object saveAndGet(Object o) {
    sink(o);
    return null;
  }

  // sink=p;Sinks;true;copyFileToDirectory;(Path,Path,CopyOption[]);;Argument[0];path-injection;df-generated
  // sink=p;Sinks;true;copyFileToDirectory;(Path,Path,CopyOption[]);;Argument[1];path-injection;df-generated
  // neutral=p;Sinks;copyFileToDirectory;(Path,Path,CopyOption[]);summary;df-generated
  public Path copyFileToDirectory(
      final Path sourceFile, final Path targetFile, final CopyOption... copyOptions)
      throws IOException {
    return Files.copy(sourceFile, targetFile, copyOptions);
  }

  // sink=p;Sinks;true;readUrl;(URL,Charset);;Argument[0];request-forgery;df-generated
  // neutral=p;Sinks;readUrl;(URL,Charset);summary;df-generated
  public String readUrl(final URL url, Charset encoding) throws IOException {
    try (InputStream in = url.openStream()) {
      byte[] bytes = in.readAllBytes();
      return new String(bytes, encoding);
    }
  }

  public static void main(String[] args) throws IOException {
    String foo = new Sinks().readUrl(new URL(args[0]), Charset.defaultCharset());
  }

  // neutral=p;Sinks;propagate;(String);summary;df-generated
  public void propagate(String s) {
    Logger logger = Logger.getLogger(Sinks.class.getSimpleName());
    logger.warning(s);
  }

  // New sink as the value of a public field is propagated to a sink.
  // sink=p;Sinks;true;fieldSink;();;Argument[this];test-sink;df-generated
  // neutral=p;Sinks;fieldSink;();summary;df-generated
  public void fieldSink() {
    sink(tainted);
  }

  // Not a new sink as this method is already defined as a manual
  // sink neutral.
  // neutral=p;Sinks;hasManualSinkNeutral;(Object);summary;df-generated
  public void hasManualSinkNeutral(Object o) {
    sink(o);
  }

  // sink=p;Sinks;true;compoundPropgate;(Sinks);;Argument[0];test-sink;df-generated
  // neutral=p;Sinks;compoundPropgate;(Sinks);summary;df-generated
  public void compoundPropgate(Sinks s) {
    s.fieldSink();
  }

  // Not a new sink because a simple type is used in an intermediate step
  // neutral=p;Sinks;wrapSinkSimpleType;(String);summary;df-generated
  public void wrapSinkSimpleType(String s) {
    Boolean b = s == "hello";
    sink(b);
  }

  // Not a new sink as this callable already has a manual sink.
  // neutral=p;Sinks;manualSinkAlreadyDefined;(Object);summary;df-generated
  public void manualSinkAlreadyDefined(Object o) {
    sink(o);
  }

  public abstract class DataWriter {
    // neutral=p;Sinks$DataWriter;write;(String);summary;df-generated
    public abstract void write(String s);
  }

  public class DataWriterKind1 extends DataWriter {
    // sink=p;Sinks$DataWriterKind1;true;write;(String);;Argument[0];test-sink;df-generated
    // neutral=p;Sinks$DataWriterKind1;write;(String);summary;df-generated
    @Override
    public void write(String s) {
      sink(s);
    }
  }

  public class DataWriterKind2 extends DataWriter {
    // sink=p;Sinks$DataWriterKind2;true;write;(String);;Argument[0];test-sink2;df-generated
    // neutral=p;Sinks$DataWriterKind2;write;(String);summary;df-generated
    @Override
    public void write(String s) {
      sink2(s);
    }
  }
}

File: csharp/ql/test/utils/modelgenerator/dataflow/Sinks.cs
using System;
using System.Web;

namespace Sinks;

public class NewSinks
{
    private string privateTainted;
    public string tainted;

    private string PrivateTaintedProp { get; set; }
    public string TaintedProp { get; set; }
    public string PrivateSetTaintedProp { get; private set; }

    // Sink defined in the extensible file next to the test.
    // neutral=Sinks;NewSinks;Sink;(System.Object);summary;df-generated
    public static void Sink(object o) => throw null;

    // Sink defined in the extensible file next to the test.
    // neutral=Sinks;NewSinks;Sink2;(System.Object);summary;df-generated
    public static void Sink2(object o) => throw null;

    // Defined as sink neutral in the file next to the neutral summary test.
    // neutral=Sinks;NewSinks;NoSink;(System.Object);summary;df-generated
    public static void NoSink(object o) => throw null;

    // Sink and Source defined in the extensible file next to the sink test.
    // sink=Sinks;NewSinks;false;SaveAndGet;(System.Object);;Argument[0];test-sink;df-generated
    // neutral=Sinks;NewSinks;SaveAndGet;(System.Object);summary;df-generated
    public static object SaveAndGet(object o)
    {
        Sink(o);
        return null;
    }

    // New sink
    // sink=Sinks;NewSinks;false;WrapResponseWrite;(System.Object);;Argument[0];html-injection;df-generated
    // neutral=Sinks;NewSinks;WrapResponseWrite;(System.Object);summary;df-generated
    public void WrapResponseWrite(object o)
    {
        var response = new HttpResponse();
        response.Write(o);
    }

    // NOT new sink as method is private
    private void PrivateWrapResponseWrite(object o)
    {
        var response = new HttpResponse();
        response.Write(o);
    }

    // New sink
    // sink=Sinks;NewSinks;false;WrapResponseWriteFile;(System.String);;Argument[0];html-injection;df-generated
    // neutral=Sinks;NewSinks;WrapResponseWriteFile;(System.String);summary;df-generated
    public void WrapResponseWriteFile(string s)
    {
        var response = new HttpResponse();
        response.WriteFile(s);
    }

    // New sink
    // sink=Sinks;NewSinks;false;WrapFieldResponseWriteFile;();;Argument[this];html-injection;df-generated
    // neutral=Sinks;NewSinks;WrapFieldResponseWriteFile;();summary;df-generated
    public void WrapFieldResponseWriteFile()
    {
        var response = new HttpResponse();
        response.WriteFile(tainted);
    }

    // NOT new sink as field is private
    // neutral=Sinks;NewSinks;WrapPrivateFieldResponseWriteFile;();summary;df-generated
    public void WrapPrivateFieldResponseWriteFile()
    {
        var response = new HttpResponse();
        response.WriteFile(privateTainted);
    }

    // New sink
    // sink=Sinks;NewSinks;false;WrapPropResponseWriteFile;();;Argument[this];html-injection;df-generated
    // neutral=Sinks;NewSinks;WrapPropResponseWriteFile;();summary;df-generated
    public void WrapPropResponseWriteFile()
    {
        var response = new HttpResponse();
        response.WriteFile(TaintedProp);
    }

    // NOT new sink as property is private
    // neutral=Sinks;NewSinks;WrapPrivatePropResponseWriteFile;();summary;df-generated
    public void WrapPrivatePropResponseWriteFile()
    {
        var response = new HttpResponse();
        response.WriteFile(PrivateTaintedProp);
    }

    // NOT new sink as property setter is private
    // neutral=Sinks;NewSinks;WrapPropPrivateSetResponseWriteFile;();summary;df-generated
    public void WrapPropPrivateSetResponseWriteFile()
    {
        var response = new HttpResponse();
        response.WriteFile(PrivateSetTaintedProp);
    }

    // Not a new sink because a simple type is used in an intermediate step
    // neutral=Sinks;NewSinks;WrapResponseWriteFileSimpleType;(System.String);summary;df-generated
    public void WrapResponseWriteFileSimpleType(string s)
    {
        var r = s == "hello";
        Sink(r);
    }

    // Not a new sink as this callable has been manually modelled
    // as sink neutral.
    // neutral=Sinks;NewSinks;ManualSinkNeutral;(System.Object);summary;df-generated
    public void ManualSinkNeutral(object o)
    {
        Sink(o);
    }

    // Not a new sink as this callable already has a manual sink.
    // neutral=Sinks;NewSinks;ManualSinkAlreadyDefined;(System.Object);summary;df-generated
    public void ManualSinkAlreadyDefined(object o)
    {
        Sink(o);
    }

    public abstract class DataWriter
    {
        // neutral=Sinks;NewSinks+DataWriter;Write;(System.Object);summary;df-generated
        public abstract void Write(object o);
    }

    public class DataWriterKind1 : DataWriter
    {
        // sink=Sinks;NewSinks+DataWriterKind1;true;Write;(System.Object);;Argument[0];test-sink;df-generated
        // neutral=Sinks;NewSinks+DataWriterKind1;Write;(System.Object);summary;df-generated
        public override void Write(object o)
        {
            Sink(o);
        }
    }

    public class DataWriterKind2 : DataWriter
    {
        // sink=Sinks;NewSinks+DataWriterKind2;true;Write;(System.Object);;Argument[0];test-sink2;df-generated
        // neutral=Sinks;NewSinks+DataWriterKind2;Write;(System.Object);summary;df-generated
        public override void Write(object o)
        {
            Sink2(o);
        }
    }
}

public class CompoundSinks
{
    // neutral=Sinks;CompoundSinks;WrapNewSinkProp;(Sinks.NewSinks);summary;df-generated
    // sink=Sinks;CompoundSinks;false;WrapNewSinkProp;(Sinks.NewSinks);;Argument[0];html-injection;df-generated
    public void WrapNewSinkProp(NewSinks ns)
    {
        ns.WrapPropResponseWriteFile();
    }

    // neutral=Sinks;CompoundSinks;WrapNewSinkField;(Sinks.NewSinks);summary;df-generated
    // sink=Sinks;CompoundSinks;false;WrapNewSinkField;(Sinks.NewSinks);;Argument[0];html-injection;df-generated
    public void WrapNewSinkField(NewSinks ns)
    {
        ns.WrapFieldResponseWriteFile();
    }
}
